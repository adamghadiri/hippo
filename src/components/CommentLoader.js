import React from "react";
import ContentLoader from "react-content-loader";

export default () => (
  <div className="list-item">
    <ContentLoader
      height="60"
      speed={2}
      primaryColor="#f3f3f3"
      secondaryColor="#e2e2e2"
    >
      <rect x="0" y="10%" rx="1" ry="1" width="25%" height="3" />
      <rect x="0" y="40%" rx="1" ry="1" width="85%" height="3" />
      <rect x="0" y="60%" rx="1" ry="1" width="80%" height="3" />
      <rect x="0" y="80%" rx="1" ry="1" width="40%" height="3" />
    </ContentLoader>
  </div>
);
