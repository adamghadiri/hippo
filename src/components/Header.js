import React from "react";
import { Link } from "react-router-dom";

export default () => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <img className="header__logo" src="/images/hacker-news-logo.jpg" />
        <Link className="header__title" to="/">
          Hacker News
        </Link>
      </div>
    </div>
  </header>
);
