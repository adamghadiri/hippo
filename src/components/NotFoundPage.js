import React from "react";
import { Link } from "react-router-dom";

const NotFoundPage = () => (
  <div className="not-found">
    <div className="not-found__item">ðŸ˜•</div>
    <div className="not-found__item">Page not found!</div>
    <Link className="not-found__button" to="/">
      Go home
    </Link>
  </div>
);

export default NotFoundPage;
