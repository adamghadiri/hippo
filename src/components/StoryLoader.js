import React from "react";
import ContentLoader from "react-content-loader";

export default () => (
  <div className="list-item">
    <ContentLoader
      height="33"
      speed={2}
      primaryColor="#f3f3f3"
      secondaryColor="#e2e2e2"
    >
      <rect x="0" y="25%" rx="1" ry="1" width="60%" height="4" />
      <rect x="0" y="75%" rx="1" ry="1" width="40%" height="3" />
    </ContentLoader>
  </div>
);
