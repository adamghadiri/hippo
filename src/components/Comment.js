import React from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import parse from "html-react-parser";
import { itemSelector } from "../selectors/items";
import actions from "../actions/items/actions";
import CommentLoader from "./CommentLoader";
import { formatTime } from "./util";

export class Comment extends React.Component {
  componentDidMount() {
    this.props.subscribeToItemById(this.props.itemId);
  }
  componentWillUnmount() {
    this.props.subscribeToItemByIdCancel(this.props.itemId);
  }
  render() {
    const { loading, data, error } = this.props.item;

    let show = () => {};
    if (data) {
      const { by, text, time } = data;
      show = () =>
        text && (
          <div className="list-item comment">
            <h6 className="comment-username">
              @{by} {formatTime(time)}
            </h6>
            <div className="comment-content">{parse(text)}</div>
          </div>
        );
    }
    return (
      <div>
        {loading && <CommentLoader />}
        {!loading && !error && show()}
        {error && (
          <div className="list-item list-item--message">
            <span>Could not load the comment. Try again later.</span>
          </div>
        )}
      </div>
    );
  }
}

const makeMapStateToProps = () => {
  return (state, props) => {
    return { item: itemSelector(state, props) };
  };
};

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      subscribeToItemById: actions.subscribeToItemById,
      subscribeToItemByIdCancel: actions.subscribeToItemByIdCancel
    },
    dispatch
  );

export default connect(
  makeMapStateToProps,
  mapDispatchToProps
)(Comment);
