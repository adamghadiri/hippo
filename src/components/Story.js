import React from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { Link } from "react-router-dom";
import parse from "html-react-parser";
import { makeStorySelector } from "../selectors/stories";
import { makeItemSelector } from "../selectors/items";
import actions from "../actions/stories/actions";
import itemActions from "../actions/items/actions";
import Comment from "./Comment";
import StoryLoader from "./StoryLoader";
import CommentLoader from "./CommentLoader";
import { formatTime } from "./util";
import uuid from "uuid/v1";

export class Story extends React.Component {
  componentDidMount() {
    if (this.props.showRank) {
      this.props.subscribeToStoryByRank(this.props.rank);
    } else if (this.props.showComments) {
      this.props.subscribeToItemById(this.props.itemId);
    }
  }
  componentWillUnmount() {
    if (this.props.showRank) {
      this.props.subscribeToStoryByRankCancel(this.props.rank);
    } else if (this.props.showComments) {
      this.props.subscribeToItemByIdCancel(this.props.itemId);
    }
  }
  render() {
    const { loading, data, error } = this.props.showRank
      ? this.props.story
      : this.props.item;

    let showStory = () => {},
      showStoryContent = () => {},
      showComments = () => {};

    if (data) {
      const { title, by, id, kids, score, time, url, text } = data;

      showStory = () => {
        const titleToShow = (
          <p className="list-item__title">
            {this.props.showRank ? this.props.rank + 1 + "." + title : title}
          </p>
        );

        const commentsLink = (
          <Link
            className="list-item__subtitle list-item__subtitle--link"
            to={`story/${id}`}
          >
            <span>top 20 comments</span>
          </Link>
        );

        const subtitle = (
          <div className="list-item__subtitle">
            {score} points by @{by} created {formatTime(time)}{" "}
            {this.props.showRank && commentsLink}
          </div>
        );

        return (
          <div className="list-item content-container">
            <div>
              {url ? (
                <div>
                  <div className="list-item__title">
                    <a href={url}>{titleToShow}</a>
                  </div>
                </div>
              ) : (
                <Link to={this.props.showRank ? `story/${id}` : `${id}`}>
                  {titleToShow}
                </Link>
              )}
            </div>
            {subtitle}
          </div>
        );
      };

      showStoryContent = () =>
        text && <div className="list-item--content">{parse(text)}</div>;

      showComments = () =>
        kids
          .slice(0, 20)
          .map(commentId => <Comment key={commentId} itemId={commentId} />);
    }

    return (
      <div>
        {loading &&
          (this.props.showComments ? (
            <div className="content-container">
              <StoryLoader />
              {[...Array(20)].map((_, i) => (
                <CommentLoader key={i} />
              ))}
            </div>
          ) : (
            <StoryLoader />
          ))}
        {!loading &&
          !error &&
          (this.props.showComments ? (
            <div className="content-container">
              {showStory()}
              {showStoryContent()}
              {showComments()}
            </div>
          ) : (
            showStory()
          ))}
        {error && (
          <div className="list-item list-item--message content-container">
            <span>Could not load the story. Try again later.</span>
          </div>
        )}
      </div>
    );
  }
}

const makeMapStateToProps = () => {
  const storySelector = makeStorySelector();
  const itemSelector = makeItemSelector();
  const mapStateToProps = (state, props) => {
    return {
      story: storySelector(state, props),
      item: itemSelector(state, props)
    };
  };
  return mapStateToProps;
};

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      subscribeToStoryByRank: actions.subscribeToStoryByRank,
      subscribeToStoryByRankCancel: actions.subscribeToStoryByRankCancel,
      subscribeToItemById: itemActions.subscribeToItemById,
      subscribeToItemByIdCancel: itemActions.subscribeToItemByIdCancel
    },
    dispatch
  );

export default connect(
  makeMapStateToProps,
  mapDispatchToProps
)(Story);
