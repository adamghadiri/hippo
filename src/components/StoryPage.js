import React from "react";
import Story from "./Story";

export default (props) => (
  <Story showComments={true} itemId={props.match.params.id} />
);
