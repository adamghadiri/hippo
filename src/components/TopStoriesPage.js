import React from "react";
import Story from "./Story";

export default () => (
  <div className="content-container">
    <h3 className="list-header">Popular updates</h3>
    <div className="list-body">
      {[...Array(10).keys()].map(rank => (
        <Story key={rank} showRank={true} rank={rank} />
      ))}
    </div>
  </div>
);
        