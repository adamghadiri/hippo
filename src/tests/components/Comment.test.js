import React from "react";
import { shallow } from "enzyme";
import { Comment } from "../../components/Comment";

const props = {
  subscribeToItemById: jest.fn(),
  subscribeToItemByIdCancel: jest.fn()
};

const comment = {
  loading: false,
  data: {
    by: "TestAuthor",
    time: 31321,
    text: "sometext"
  },
  error: null
};

test("should render Comment by id correctly", () => {
  const wrapper = shallow(
    <Comment itemId="test-item-id" item={comment} {...props} />
  );
  expect(wrapper).toMatchSnapshot();
});
