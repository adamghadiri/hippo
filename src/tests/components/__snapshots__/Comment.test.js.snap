// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Comment by id correctly 1`] = `
<div>
  <div
    className="list-item comment"
  >
    <h6
      className="comment-username"
    >
      @
      TestAuthor
       
      03:42:01 1970/01/01
    </h6>
    <div
      className="comment-content"
    >
      sometext
    </div>
  </div>
</div>
`;
